### PR TITLE
MGDOBR-658: Add bridges filtering support

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-658](https://issues.redhat.com/browse/MGDOBR-658)

This is a duplication of #741 but hopefully this will get a proper SonarCloud analysis.

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x ] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

**Comment by manstis** [comment]
@jomarko @danielezonca I am at a loss now.

This PR contains a single commit and was created after (and includes) the GHA workflow change for `fetch-depth: 0`.

SonarCloud still steadfastly refuses to see the [unit tests](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/752/files#diff-23819953a92802e4d712ad1b898ab58b1080b6516ccab28f16c513a3c86571b6R113-R209).

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 1]
```diff
@@ -16,6 +19,9 @@
 @Transactional
 public class BridgeDAO implements PanacheRepositoryBase<Bridge, String> {
 
+    private static final int FILTER_NAME = 1;
```
Comment (lucamolteni) [inline manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:22]: `FILTER_BY_NAME`

**Review by lucamolteni — CHANGES_REQUESTED** [review]

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 2]
```diff
@@ -34,10 +40,41 @@ public Bridge findByIdAndCustomerId(String id, String customerId) {
...
+    public ListResult<Bridge> findByCustomerId(String customerId, QueryResourceInfo queryInfo) {
         Parameters parameters = Parameters.with("customerId", customerId);
-        long total = find("#BRIDGE.findByCustomerId", parameters).count();
-        List<Bridge> bridges = find("#BRIDGE.findByCustomerId", parameters).page(queryInfo.getPageNumber(), queryInfo.getPageSize()).list();
+        long total;
+        List<Bridge> bridges;
+        String filterName = queryInfo.getFilterInfo().getFilterName();
+        Set<ManagedResourceStatus> filterStatus = queryInfo.getFilterInfo().getFilterStatus();
+
+        int query = 0;
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:50]: Done.

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java (hunk 1 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:22]: Done.

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 3]
```diff
@@ -16,6 +19,9 @@
 @Transactional
 public class BridgeDAO implements PanacheRepositoryBase<Bridge, String> {
 
+    private static final int FILTER_NAME = 1;
+    private static final int FILTER_STATUS = 2;
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:23]: Done.

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 4]
```diff
@@ -34,10 +40,41 @@ public Bridge findByIdAndCustomerId(String id, String customerId) {
...
+            case FILTER_STATUS:
+                total = find("#BRIDGE.findByCustomerIdFilterByStatus", parameters).count();
+                bridges = find("#BRIDGE.findByCustomerIdFilterByStatus", parameters).page(queryInfo.getPageNumber(), queryInfo.getPageSize()).list();
+                break;
+            case (FILTER_NAME + FILTER_STATUS):
+                total = find("#BRIDGE.findByCustomerIdFilterByNameAndStatus", parameters).count();
+                bridges = find("#BRIDGE.findByCustomerIdFilterByNameAndStatus", parameters).page(queryInfo.getPageNumber(), queryInfo.getPageSize()).list();
+                break;
+            default:
+                total = find("#BRIDGE.findByCustomerId", parameters).count();
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:74]: @lucamolteni Thank-you for the pointer. I investigated Panache as you proposed and it actually rewrites the `SELECT ... FROM ...` statement into a `SELECT COUNT(*) FROM ...` statement in `io.quarkus.panache.hibernate.common.runtime.PanacheJpaUtil#getCountQuery(..)` called from the `count()` method. Therefore this is actually as efficient as I could hand-craft with additional queries.

[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:78]: I'll have a look at Hibernate Filters though, which could improve the whole situation too.

I didn't know such a thing existed to be honest but whilst looking through the Panache code I found them.

[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java]: @lucamolteni Filtering has been completely refactored and looks much better for it :dancing_women:

**Comment by manstis** [comment]
@jomarko @lucamolteni @Srihari1192 This PR has been refactored to leverage Hibernate better for filtering.

Please feel free to re-review.

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 5]
```diff
@@ -34,10 +38,22 @@ public Bridge findByIdAndCustomerId(String id, String customerId) {
...
     }
 
-    public ListResult<Bridge> findByCustomerId(String customerId, QueryInfo queryInfo) {
+    public ListResult<Bridge> findByCustomerId(String customerId, QueryResourceInfo queryInfo) {
         Parameters parameters = Parameters.with("customerId", customerId);
-        long total = find("#BRIDGE.findByCustomerId", parameters).count();
-        List<Bridge> bridges = find("#BRIDGE.findByCustomerId", parameters).page(queryInfo.getPageNumber(), queryInfo.getPageSize()).list();
+        PanacheQuery<Bridge> query = find("#BRIDGE.findByCustomerId", parameters);
+
+        String filterName = queryInfo.getFilterInfo().getFilterName();
```
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:45]: Look how pretty the new filtering code is!

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/EnumTypeManagedResourceStatus.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.redhat.service.smartevents.infra.models.dto.ManagedResourceStatus;
+
+/**
+ * Custom Type for ManagedResourceStatus enumeration.
+ * If we define the type of the applicable {@link FilterDef} as the enum class itself Hibernate uses Java
+ * serialization to a PostgreSQL bytea type as the value in Filters. This does not work since we choose to explicitly
+ * store Enums as {@link javax.persistence.EnumType#STRING}. This class therefore represents a custom type usable in Filters.
+ */
+public class EnumTypeManagedResourceStatus extends EnumType<ManagedResourceStatus> {
```
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/dao/EnumTypeManagedResourceStatus.java:18]: Look how less pretty this hack-around is.. but I think it's a worthy alternative.

Without this we need to convert the `ManagedResourceStatus` enum values to `String` to use in the `FilterDef`.

File: manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java [hunk 7]
```diff
@@ -34,10 +38,22 @@ public Bridge findByIdAndCustomerId(String id, String customerId) {
...
+        Set<ManagedResourceStatus> filterStatus = queryInfo.getFilterInfo().getFilterStatus();
+        if (Objects.nonNull(filterName)) {
+            query.filter("byName", Parameters.with("name", filterName + "%"));
+        }
+        if (Objects.nonNull(filterStatus) && !filterStatus.isEmpty()) {
+            query.filter("byStatus", Parameters.with("status", filterStatus));
```
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/dao/BridgeDAO.java:51]: Without the hackery magic (see below) I'd need to convert the statuses to `String`.

**Review by kelvah — APPROVED** [review]

**Review by Srihari1192 — APPROVED** [review]

**Comment by manstis** [comment]
@jomarko @r00ta Could you please review (or say you don't want to/feel the need to).

**Comment by sutaakar** [comment]
Jozef is on PTO, removing him from reviewers

**Review by lucamolteni — APPROVED** [review]

**Pull request merged**